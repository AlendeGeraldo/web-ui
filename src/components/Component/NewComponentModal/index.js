import React from 'react';
import PropTypes from 'prop-types';
import { Modal, Form, Input } from 'antd';

const NewComponentModal = (props) => {
  const { visible, form, history } = props;
  const { onCreate, onCancel } = props;
  const { getFieldDecorator, getFieldsError } = form;

  /* 
  This function is used to check if form has errors
*/
  const hasErrors = (fieldsError) => {
    return Object.keys(fieldsError).some((field) => fieldsError[field]);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    form.validateFields(async (err, values) => {
      if (err) {
        return;
      }
      onCreate(values.name, history);
    });
  };

  return (
    <Modal
      visible={visible}
      title='Novo Componente'
      okText='Criar'
      cancelText='Cancelar'
      onCancel={onCancel}
      onOk={handleSubmit}
      okButtonProps={{ disabled: hasErrors(getFieldsError()) }}
    >
      <Form layout='vertical'>
        <Form.Item label='Qual o nome do seu componente?'>
          {getFieldDecorator('name', {
            rules: [
              {
                required: true,
                message: 'Por favor insira um nome para o componente!',
              },
            ],
          })(<Input allowClear autoFocus />)}
        </Form.Item>
      </Form>
    </Modal>
  );
};

NewComponentModal.propTypes = {
  visible: PropTypes.bool.isRequired,
  onCancel: PropTypes.func.isRequired,
  onCreate: PropTypes.func.isRequired,
  form: PropTypes.shape({
    getFieldDecorator: PropTypes.func,
    getFieldsError: PropTypes.func,
  }).isRequired,
};

export default Form.create({ name: 'new-component' })(NewComponentModal);
