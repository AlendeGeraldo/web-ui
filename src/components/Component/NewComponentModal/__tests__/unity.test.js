import React from 'react';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { Modal } from 'antd';
import { shallow, mount } from 'enzyme';
import configureMockStore from 'redux-mock-store';

import NewComponentModal from '..';

const initialState = {
  components: {
    componentList: [],
    modalIsVisible: false,
    loading: false,
    error: null,
  },
};

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);
const store = mockStore(initialState);

jest.mock('react-router-dom', () => ({
  useHistory: () => ({
    push: jest.fn(),
  }),
}));

describe('NewComponentModal component', () => {
  it('is expected render without crashing', () => {
    shallow(
      <Provider store={store}>
        <NewComponentModal />
      </Provider>
    );
  });

  it('is expected render html correctly', () => {
    const wrapper = mount(
      <Provider store={store}>
        <NewComponentModal />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });

  it('should render self and subcomponents', () => {
    const wrapper = mount(
      <Provider store={store}>
        <NewComponentModal />
      </Provider>
    );

    expect(wrapper.find('Form(NewComponentModal)').exists()).toBeTruthy();
    expect(wrapper.find(NewComponentModal).exists()).toBeTruthy();
    expect(wrapper.find(Modal).exists()).toBeTruthy();
  });

  it('onOk should be called', () => {
    const wrapper = mount(
      <Provider store={store}>
        <NewComponentModal />
      </Provider>
    );

    const event = { preventDefault: () => {} };
    const modal = wrapper.find(Modal).instance();
    const spyHandleOk = jest.spyOn(modal, 'handleOk');
    modal.handleOk(event);
    expect(spyHandleOk).toHaveBeenCalled();
  });

  it('onCancel should be called', () => {
    const wrapper = mount(
      <Provider store={store}>
        <NewComponentModal />
      </Provider>
    );

    const modal = wrapper.find(Modal).instance();
    const spyHandleCancel = jest.spyOn(modal, 'handleCancel');
    modal.handleCancel();
    expect(spyHandleCancel).toHaveBeenCalled();
  });
});
