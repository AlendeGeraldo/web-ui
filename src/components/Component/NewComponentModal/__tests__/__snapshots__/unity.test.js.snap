// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewComponentModal component is expected render html correctly 1`] = `
<Modal
  cancelText="Cancelar"
  confirmLoading={false}
  maskTransitionName="fade"
  okButtonProps={
    Object {
      "disabled": false,
    }
  }
  okText="Criar"
  okType="primary"
  onCancel={[Function]}
  onOk={[Function]}
  title="Novo Componente"
  transitionName="zoom"
  visible={false}
  width={520}
>
  <Form
    colon={true}
    hideRequiredMark={false}
    layout="vertical"
    onSubmit={[Function]}
  >
    <FormItem
      hasFeedback={false}
      label="Qual o nome do seu componente?"
    >
      <Input
        allowClear={true}
        autoFocus={true}
        data-__field={
          Object {
            "name": "name",
          }
        }
        data-__meta={
          Object {
            "name": "name",
            "originalProps": Object {
              "allowClear": true,
              "autoFocus": true,
              "type": "text",
            },
            "ref": null,
            "rules": Array [
              Object {
                "message": "Por favor insira um nome para o componente!",
                "required": true,
              },
            ],
            "trigger": "onChange",
            "validate": Array [
              Object {
                "rules": Array [
                  Object {
                    "message": "Por favor insira um nome para o componente!",
                    "required": true,
                  },
                ],
                "trigger": Array [
                  "onChange",
                ],
              },
            ],
            "valuePropName": "value",
          }
        }
        id="new-component_name"
        onChange={[Function]}
        type="text"
      />
    </FormItem>
  </Form>
</Modal>
`;
